Modules/selectmodule.c

```diff
@@ -257,26 +272,43 @@ select_select_impl(PyObject *module, PyObject *rlist, PyObject *wlist,
                    PyObject *xlist, PyObject *timeout_obj)
 /*[clinic end generated code: output=2b3cfa824f7ae4cf input=e467f5d68033de00]*/
 {
-#ifdef SELECT_USES_HEAP
+    size_t setsize;
     pylist *rfd2obj, *wfd2obj, *efd2obj;
-#else  /* !SELECT_USES_HEAP */
-    /* XXX: All this should probably be implemented as follows:
-     * - find the highest descriptor we're interested in
-     * - add one
-     * - that's the size
-     * See: Stevens, APitUE, $12.5.1
-     */
-    pylist rfd2obj[FD_SETSIZE + 1];
-    pylist wfd2obj[FD_SETSIZE + 1];
-    pylist efd2obj[FD_SETSIZE + 1];
-#endif /* SELECT_USES_HEAP */
     PyObject *ret = NULL;
-    fd_set ifdset, ofdset, efdset;
+    fd_set *ifdset, *ofdset, *efdset;
+#if !defined(MS_WINDOWS)
+    fd_set ifdset_stack, ofdset_stack, efdset_stack;
+#if !defined(SELECT_USES_HEAP)
+    pylist rfd2obj_stack[FD_SETSIZE + 1];
+    pylist wfd2obj_stack[FD_SETSIZE + 1];
+    pylist efd2obj_stack[FD_SETSIZE + 1];
+#endif /* !defined(SELECT_USES_HEAP) */
+#else
+    Py_ssize_t rsize, wsize, xsize;
+#endif /* !defined(MS_WINDOWS) */
     struct timeval tv, *tvp;
     int imax, omax, emax, max;
     int n;
     _PyTime_t timeout, deadline = 0;
 
+    setsize = FD_SETSIZE;
+#if defined(MS_WINDOWS)
+    /* On windows we allocated the fd sets dynamically
+    based on the inputs size */
+    rsize = PySequence_Length(rlist);
+    wsize = PySequence_Length(wlist);
+    xsize = PySequence_Length(xlist);
+
+    if (rsize < 0 || wsize < 0 || xsize < 0) {
+        PyErr_SetString(PyExc_ValueError, "arguments 1-3 must be sequences");
+        return NULL;
+    }
+
+    if ((size_t)rsize > setsize) setsize = (size_t)rsize;
```

Comment:
You may write: ``setsize = Py_MAX(setsize, (size_t)rsize);`` (same for the 2 lines below). IMO it's more explicit (and how I would write it in Python).